Call node destructors when freeing the Context.

nweston opened the pull request:
Adds a subclass of BumpPtrAllocatorImpl which only allocates a single
type, allowing it to walk the slabs and destroy objects during its own
destructor.

Context now has a separate allocator for each node type, with a
templated Allocate method which is specialized to use the correct
allocator.

The `Create` methods now call `Context::Allocate` explicitly, and use the built-in placement new rather than our own version. So this code:
```c++
return new (C) Module(C);
```
becomes this:
```c++
return new (C.Allocate<Module>()) Module(C);
```
This allows us to pass the type information down to the `Context`, so it can pick the right allocator.


There are two things which I don't really like about this design:

1. `DtorBumpPtrAllocator` is very dependent on the details of `BumpPtrAllocatorImpl`. But the public API is insufficient for our purposes, and it seems better to use an existing allocator which is presumably well-tested than to start from scratch. And in any event we aren't likely to make a lot of changes to this code. So I think this is ultimately the right way to go.

2. `DtorBumpPtrAllocator` is responsible for calling the destructors, but the various `Create` methods determine the type of the allocated object. So essentially, the two sides of the process are handled at different layers. One can imagine a copy-paste error leading to code like:
```c++
return new (C.Allocate<Module>()) MyNode(C);
```

This won't be detected at compile time and could lead to subtle errors. But this seems to be an unavoidable consequence of the way placement new works.

AaronBallman commented on include/gtirb/Allocator.hpp:
```diff
@@ -264,6 +264,44 @@ class BumpPtrAllocatorImpl {
 
+/// Allocate memory for a specific type in an ever-growing pool, and call
+/// destructors for all allocated objects when freeing the pool.
+template <class T, size_t SlabSize = 4096, size_t SizeThreshold = SlabSize>
```
Should we make the `SlabSize` a multiple of `sizeof(T)`?

AaronBallman commented on include/gtirb/Allocator.hpp:
```diff
@@ -264,6 +264,44 @@ class BumpPtrAllocatorImpl {
+/// Allocate memory for a specific type in an ever-growing pool, and call
+/// destructors for all allocated objects when freeing the pool.
+template <class T, size_t SlabSize = 4096, size_t SizeThreshold = SlabSize>
+class DtorBumpPtrAllocator : BumpPtrAllocatorImpl<SlabSize, SizeThreshold> {
```
The way that I'm used to seeing this done is more like `SpecificBumpPtrAllocator` (https://llvm.org/viewvc/llvm-project/llvm/trunk/include/llvm/Support/Allocator.h?view=markup#l415) where it composes rather than inherits. Did you consider such a design?

nweston replied:
Oh, I wish I'd know that existed. I should probably throw out my allocator and use that instead, since it goes along with the `BumpPtrAllocator` that we've already copied.

AaronBallman commented on src/Context.cpp:
```diff
@@ -13,10 +13,23 @@
 
 using namespace gtirb;
 
+// By moving these declarations here, we avoid instantiating the default
```
Did you test this with MSVC as well as Clang?

nweston replied:
Yes.

nweston commented:
I've replaced my allocator with `SpecificBumpPtrAllocator` from LLVM, which resolves all the comments about the specifics of the allocator. And I reworked the `Create` methods as we discussed yesterday.

AaronBallman commented on include/gtirb/Allocator.hpp:
```diff
@@ -201,7 +206,7 @@ class BumpPtrAllocatorImpl {
 
   void setRedZoneSize(size_t NewSize) { RedZoneSize = NewSize; }
 
-private:
```
Is this change still needed?

nweston replied:
No.

AaronBallman commented on include/gtirb/Context.hpp:
```diff
@@ -64,17 +81,21 @@ class GTIRB_EXPORT_API Context {
-  }
+  /// \return The newly allocated memory, suitably sized for the given
+  /// type. Will return nullptr if the allocation cannot be honored.
+  template <class T> void* Allocate() const;
```
Can this be made private?

nweston replied:
Yes.

AaronBallman commented on include/gtirb/Symbol.hpp:
```diff
@@ -217,7 +217,9 @@ class GTIRB_EXPORT_API Symbol : public Node {
   /// \return The newly created object.
-  static Symbol* Create(Context& C) { return new (C) Symbol(C); }
+  static Symbol* Create(Context& C) {
+    return new (C.Allocate<Symbol>()) Symbol(C);
```
Shouldn't this be using `C.Create<Symbol>()`? (Same below.)

nweston commented:
All fixed now.
